core/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java

```diff
@@ -0,0 +1,276 @@
+        Path tempDir = createTempDir();
+        Store store = newStore(tempDir);
+        AtomicBoolean failedEngine = new AtomicBoolean(false);
+        RecoverySourceHandler handler = new RecoverySourceHandler(null, request, recoverySettings, null, logger) {
+            @Override
+            protected void failEngine(IOException cause) {
+                assertFalse(failedEngine.get());
+                failedEngine.set(true);
+            }
+        };
+        Directory dir = store.directory();
+        RandomIndexWriter writer = new RandomIndexWriter(random(), dir, newIndexWriterConfig());
+        int numDocs = randomIntBetween(10, 100);
+        for (int i = 0; i < numDocs; i++) {
+            Document document = new Document();
+            document.add(new StringField("id", Integer.toString(i), Field.Store.YES));
+            document.add(newField("field", randomUnicodeOfCodepointLengthBetween(1, 10), TextField.TYPE_STORED));
+            writer.addDocument(document);
+        }
+        writer.commit();
+        writer.close();
+
+        Store.MetadataSnapshot metadata = store.getMetadata();
+        List<StoreFileMetaData> metas = new ArrayList<>();
+        for (StoreFileMetaData md : metadata) {
+            metas.add(md);
+        }
+        final boolean throwCorruptedIndexException = randomBoolean();
+        Store targetStore = newStore(createTempDir());
+        try {
+            handler.sendFiles(store, metas.toArray(new StoreFileMetaData[0]), (md) -> {
+                if (throwCorruptedIndexException) {
+                    throw new RuntimeException(new CorruptIndexException("foo", "bar"));
+                } else {
+                    throw new RuntimeException("boom");
+                }
+            });
+            fail("exception index");
+        } catch (RuntimeException ex) {
+            assertNull(ExceptionsHelper.unwrapCorruption(ex));
+            if (throwCorruptedIndexException) {
+                assertEquals(ex.getMessage(), "[File corruption occurred on recovery but checksums are ok]");
+            } else {
+                assertEquals(ex.getMessage(), "boom");
+            }
+        } catch (CorruptIndexException ex) {
+            fail("not expected here");
+        }
+        assertFalse(failedEngine.get());
+        IOUtils.close(store, targetStore, recoverySettings);
+    }
+
+    private Store newStore(Path path) throws IOException {
+        DirectoryService directoryService = new DirectoryService(shardId, Settings.EMPTY) {
+            @Override
+            public long throttleTimeInNanos() {
+                return 0;
+            }
+
+            @Override
+            public Directory newDirectory() throws IOException {
+                return RecoverySourceHandlerTests.newFSDirectory(path);
+            }
+        };
+        return new Store(shardId, Settings.EMPTY, directoryService, new DummyShardLock(shardId));
+    }
+
+    public void corruptFile(Path[] files) throws IOException {
```

Comment:
I think we have this code now in many places, we should factor it out into `ESTestCase` or a test helper or something.